# Weather

This notebook is part of the **Computer Infrastructure** module **(Task 8)**.


***

## Tasks:

### Task 1: Create Directory Structure

Create the "Data" directory using `mkdir` command:

![Task1-1](img/task1-1.png)

Navigate into the "Data" directory using `cd` command:

![Task1-2](img/task1-2.png)

Create the "timestamps" and "weather" directories using `mkdir` command:

![Task1-3](img/task1-3.png)

Alternatively, use the "-p" option added to the `mkdir` command to create the directories and parent directory at the same time:

![Task1-4](img/task1-4.png)

### Task 2: Timestamps

Navigate into the "data/timestamps" directory using `cd` command:

![Task2-1](img/task2-1.png)

Append the current date and time to "now.txt" ten times using `date` command and `>>` operator:

![Task2-2](img/task2-2.png)

Verify the contents of "now.txt" using `more` command:

![Task2-3](img/task2-3.png)


### Task 3: Formatting Timestamps

Navigate into the "data/timestamps" directory using `cd` command:

![Task3-1](img/task2-1.png)

Append the current date and time in the `YYYYmmdd_HHMMSS` format to "formatted.txt" using `date +"format"` command and `>>` operator:

![Task3-2](img/task3-2.png)

Verify the contents of "formatted.txt" using `more` command:

![Task3-3](img/task3-3.png)


### Task 4: Create Timestamped Files

Navigate into the "data/timestamps" directory using `cd` command:

![Task4-1](img/task2-1.png)

Create an empty file with a name in the "YYYYmmdd_HHMMSS.txt" format using `touch` , ` `` ` and ``date +"%Y%m%d_%H%M%S"``commands:

![Task4-2](img/task4-2.png)

Verify that the file has been created using `ls` command:

![Task4-3](img/task4-3.png)



### Task 5: Download Today's Weather Data

Navigate into the "data/weather" directory using `cd` command:

![Task5-1](img/task5-1.png)

Download the weather data using the `wget` command, and use the `-O <filename>` option to save the file with a specific name:

![Task5-2](img/task5-2.png)

Verify that the file has been downloaded/created using `ls` command:

![Task5-3](img/task5-3.png)

### Task 6: Timestamp the Data

Navigate into the "data/weather" directory using `cd` command:

![Task6-1](img/task5-1.png)

Download the weather data using the `wget` command, and use the ` `` ` and ``date +"%Y%m%d_%H%M%S"`` commands to save the file with a specific name/date format:

![Task6-2](img/task6-2.png)

Verify that the file has been downloaded/created using `ls` command:

![Task6-3](img/task6-3.png)

### Task 7: Write the Script

Navigate into the "repository root" directory using `cd` command:

![Task7-1](img/task7-1.png)

Create an empty file using `touch` command:

![Task7-2](img/task7-2.png)

Opening the file in nano text editor using `nano` command:

![Task7-3](img/task7-3.png)

Writing the bash script using nano text editor:

**Ctrl + O** to Save

**Ctrl + x** to Exit

![Task7-4](img/task7-4.png)

Make the script executable using `chmod` command using the `+x` option to add execute permission as a program:

![Task7-5](img/task7-5.png)

Test the script by running it:

![Task7-6](img/task7-6.png)

Verify that the file has been downloaded/created in the correct directory using `cd` & `ls` commands:

![Task7-7](img/task7-7.png)

***

## Commands Used:

`mkdir`

**Creates a directory.**

Example: `mkdir data` creates a new directory called `data`.

`cd`

**Changes the current directory.**

Example: cd `data/timestamps` moves you into the `data/timestamps` directory.

`ls`

**Lists directory contents.**

Example: `ls` shows files and directories in the current directory.

`date`

**Displays the current date and time.**

Example: `date` prints the current date and time.

`more`

**Displays text files one screen at a time.**

Example: `more now.txt` allows you to view `now.txt` one screen at a time.

`>>` (right shift operator):

**Appends output to a file.**

Example: `echo "Hello" >> file.txt appends "Hello" to file.txt.

`man + command`

Displays the manual for the command.

Example: man date shows detailed documentation for the date command.

` `` ` (backticks):

**Executes commands and returns their output.**

Example: `date >> now.txt` includes the current date in the `now.txt` file.

`nano`

**Opens the Nano text editor.**

Example: `nano weather.sh` opens `weather.sh` for editing in Nano.

`touch`

**Creates an empty file or updates the timestamp of an existing file.**

Example: `touch weather.sh` creates `weather.sh` if it doesn't exist or updates its timestamp if it does.

`chmod`

**Changes file permissions.**

Example: chmod +x weather.sh makes weather.sh executable.



***

## Task 9 Pandas:

### Code:

In [20]:
import pandas as pd
from tabulate import tabulate

In [22]:
# Define the path to the JSON file
file_path = 'data/weather/20241124_004545.json' 

# Load the JSON data
weather_data = pd.read_json(file_path)

# Check if the data was loaded successfully
if not weather_data.empty:
    # Print the DataFrame in a table format
    print(tabulate(weather_data, headers='keys', tablefmt='psql'))
else:
    print("Failed to load weather data.")

+----+---------+---------------+----------+----------------------+---------------+-------------+------------+-------------------------+-----------------+------------+------------+------------+-----------+---------------------+--------------+
|    | name    |   temperature | symbol   | weatherDescription   | text          |   windSpeed |   windGust | cardinalWindDirection   |   windDirection |   humidity |   rainfall |   pressure | dayName   | date                | reportTime   |
|----+---------+---------------+----------+----------------------+---------------+-------------+------------+-------------------------+-----------------+------------+------------+------------+-----------+---------------------+--------------|
|  0 | Athenry |            10 | 05n      | Rain showers         | "Rain shower" |          35 |         61 | SW                      |             225 |         90 |        0.4 |        976 | Sunday    | 2024-11-24 00:00:00 | 00:00        |
+----+---------+---------------+

### Report:

The dataset contains weather information for a specific location, **Athenry**, recorded at a particular time. Here’s a brief explanation of the key components of the dataset:

- **Name:** The name of the location (Athenry).
- **Temperature:** The current temperature in degrees Celsius (10°C).
- **Symbol:** A code representing the weather condition (e.g., "05n" for nighttime rain).
- **WeatherDescription:** A textual description of the weather condition (e.g., "Rain showers").
- **Text:** A more detailed description of the weather (e.g., "Rain shower").
- **WindSpeed:** The speed of the wind in kilometers per hour (35 km/h).
- **WindGust:** The maximum wind gust speed recorded (61 km/h).
- **CardinalWindDirection:** The direction from which the wind is blowing (e.g., "SW" for southwest).
- **WindDirection:** The wind direction in degrees (225°).
- **Humidity:** The relative humidity percentage (90%).
- **Rainfall:** The amount of rainfall recorded (0.4 mm).
- **Pressure:** The atmospheric pressure in hPa (976 hPa).
- **DayName:** The name of the day of the week (Sunday).
- **Date:** The date of the weather report (2024-11-24).
- **ReportTime:** The time at which the weather data was recorded (00:00).

**Overall**, this dataset provides a snapshot of the weather conditions in Athenry, including temperature, wind, humidity, and atmospheric pressure, at a specific date and time.

***

## Resources:

**Markdown:**

https://www.jetbrains.com/help/hub/markdown-syntax.html#quick-notes-markdown-characters

**Bash:**

https://devhints.io/bash

**Command Line:**

https://www.codecademy.com/article/command-line-commands 

https://ubuntu.com/tutorials/command-line-for-beginners#1-overview

**Codespaces:**

https://docs.github.com/en/codespaces/getting-started/quickstart

***

## End